### PR TITLE
feat: add regret to the project

### PR DESCRIPTION
**Pull request by furadnik**

### CODE REVIEW

File: incomplete_cooperative/regret.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        self.coalitions_to_player_ids = get_coalition_player_id_map(number_of_players)
+        self.regret = np.zeros((self.number_of_regret_minimizers, self.number_of_coalitions), dtype=RMValue)
+
+    def get_metacoalition_id(self, metacoalition: Iterable[Coalition]) -> int:
+        """Properly get an id of a metacoalition."""
+        coalition_player_ids = self.coalitions_to_player_ids[
+            [x.id for x in metacoalition if len(x) not in [0, 1, 2**self.number_of_players]]]
+        return np.sum(2 ** coalition_player_ids)
+
+    def get_actions_distribution(self, past_actions: Iterable[Coalition] | int) -> np.ndarray:
```
Comment (DavidSych) [inline incomplete_cooperative/regret.py:111]: Maybe call this `regret_matching` and add `predictive` and `plus` boolean arguments.

File: incomplete_cooperative/regret.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        self.viable_metacoalitions = len(self.meta_rank_to_id)
+        # map metacoalition id -> metacoalition size rank
+        self.meta_id_to_rank = np.zeros(self.viable_metacoalitions, dtype=int)
+        self.meta_id_to_rank[self.meta_rank_to_id] = np.arange(self.viable_metacoalitions)
+
+        # regret minimizer for every internal node -- except for the bottom layer of game tree
+        self.number_of_regret_minimizers = coalitions_up_to(self.number_of_coalitions, limit_of_revealed - 1)
+
+        self.coalitions_to_player_ids = get_coalition_player_id_map(number_of_players)
+        self.regret = np.zeros((self.number_of_regret_minimizers, self.number_of_coalitions), dtype=RMValue)
```
Comment (DavidSych) [inline incomplete_cooperative/regret.py:103]: I would call it `cumulative_regret` and also keep track of `cumulative_strategy`.

File: incomplete_cooperative/regret.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        experienced_losses = np.zeros(self.viable_metacoalitions, dtype=RMValue)
+        experienced_losses[used_metacoalition_ranks] = bottom_losses
+        for i in range(self.number_of_regret_minimizers - 1, -1, -1):
+            metacoalition = self.meta_rank_to_id[i]
+            coalition_pids = list(Coalition(metacoalition).players)
+            next_coalition_pids = [x for x in range(self.number_of_coalitions) if x not in coalition_pids]
+            next_metacoalitions = [Coalition.from_players(coalition_pids + [x]).id for x in next_coalition_pids]
+            next_metacoalition_ranks = self.meta_id_to_rank[next_metacoalitions]
+            losses[i, next_coalition_pids] = experienced_losses[next_metacoalition_ranks]
+            experienced_losses[i] = (losses[i] * self.get_actions_distribution(int(metacoalition))).sum()
```
Comment (DavidSych) [inline incomplete_cooperative/regret.py:144]:
```
metacoalition_strategy = self.get_actions_distribution(int(metacoalition))  # To be called `regret_matching`
experienced_losses[i] = (losses[i] * metacoalition_strategy).sum()
weight = self.iteration if plus else 1  # The average strategy uses linear averaging
self.cumulative_strategy[metacoalition] += weight * metacoalition_strategy
```

File: incomplete_cooperative/regret.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    """
+
+    def __init__(self, number_of_players: int, limit_of_revealed: int) -> None:
+        """Initialize RM.
+
+        Arguments:
+            number_of_players: number of players in the games.
+            limit_of_revealed: maximum number of coalitions to reveal.
+        """
+        self.number_of_players = number_of_players
```
Comment (DavidSych) [inline incomplete_cooperative/regret.py:86]: `self.iteration = 0`

File: incomplete_cooperative/regret.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+        for i in range(self.number_of_regret_minimizers - 1, -1, -1):
+            metacoalition = self.meta_rank_to_id[i]
+            coalition_pids = list(Coalition(metacoalition).players)
+            next_coalition_pids = [x for x in range(self.number_of_coalitions) if x not in coalition_pids]
+            next_metacoalitions = [Coalition.from_players(coalition_pids + [x]).id for x in next_coalition_pids]
+            next_metacoalition_ranks = self.meta_id_to_rank[next_metacoalitions]
+            losses[i, next_coalition_pids] = experienced_losses[next_metacoalition_ranks]
+            experienced_losses[i] = (losses[i] * self.get_actions_distribution(int(metacoalition))).sum()
+
+        self.regret += losses[np.arange(self.number_of_regret_minimizers)] - experienced_losses[np.arange(self.number_of_regret_minimizers), np.newaxis]
```
Comment (DavidSych) [inline incomplete_cooperative/regret.py:146]: self.iteration += 1

File: incomplete_cooperative/regret.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+        """Scan every possible action sequence, and update the regret of all the regret minimizers.
+
+        Arguments:
+            bottom_losses: the loss of the bottom layer of the game tree.
+                    this should be an array of size (number_of_coalitions,).
+            used_actions: the list of list of actions used to generate the bottom_losses.
+        """
+        used_metacoalition_ranks = [self.meta_id_to_rank[self.get_metacoalition_id(x)]
+                                    for x in used_actions]
+        losses = np.zeros((self.number_of_regret_minimizers, self.number_of_coalitions),
```
Comment (DavidSych) [inline incomplete_cooperative/regret.py:133]: `meta_q_values` instead of `loss`?

File: incomplete_cooperative/regret.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+                          dtype=RMValue)
+        experienced_losses = np.zeros(self.viable_metacoalitions, dtype=RMValue)
```
Comment (DavidSych) [inline incomplete_cooperative/regret.py:135]: `meta_values` instead of `experienced_losses`?

File: incomplete_cooperative/regret.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+        number_of_coalitions: number of coalitions which aren't in K_0.
+        limit_of_revealed: maximum number of coalitions to reveal, i.e. maximum size of metacoal.
+        meta_rank_to_id: map metacoalition size rank -> metacoalition id.
+        meta_id_to_rank: map metacoalition id -> metacoalition size rank.
+        regret: an array of instant regrets of each regret minimizer (zero if an action is inapplicable).
+            This array has size (number_of_regret_minimizers, number_of_coalitions).
+            It is indexed by the meta-coalition ids.
+    """
+
+    def __init__(self, number_of_players: int, limit_of_revealed: int) -> None:
```
Comment (DavidSych) [inline incomplete_cooperative/regret.py:79]: also needs `plus` and `predictive` boolean arguments.

File: incomplete_cooperative/regret.py [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+        If past_actions is an int, it is considered a metacoalition id.
+        Otherwise, it is considered as a list of coalitions of the original game.
+        """
+        metacoalition = past_actions if isinstance(past_actions, int) else self.get_metacoalition_id(past_actions)
+        rank = self.meta_id_to_rank[metacoalition]
+        if sum(self.regret[rank]) == 0:
+            return np.ones(self.number_of_coalitions) / self.number_of_coalitions  # TODO: not all viable!
+        return self.regret[rank] * (self.regret[rank] > 0) / sum(self.regret[rank] * self.regret[rank] > 0)
+
+    def apply_regret(self, bottom_losses: np.ndarray, used_actions: list[list[Coalition]]) -> None:
```
Comment (DavidSych) [inline incomplete_cooperative/regret.py:123]: `regret_minimization_iteration`?

File: incomplete_cooperative/regret.py [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+        """Scan every possible action sequence, and update the regret of all the regret minimizers.
+
+        Arguments:
+            bottom_losses: the loss of the bottom layer of the game tree.
```
Comment (DavidSych) [inline incomplete_cooperative/regret.py:127]: `terminal_losses`?

File: incomplete_cooperative/regret.py (hunk 2 above)
Comment (furadnik) [reply incomplete_cooperative/regret.py:103]: Done.

File: incomplete_cooperative/regret.py (hunk 10 above)
Comment (furadnik) [reply incomplete_cooperative/regret.py:128]: Changed.

File: incomplete_cooperative/regret.py (hunk 9 above)
Comment (furadnik) [reply incomplete_cooperative/regret.py:124]: Changed.

File: incomplete_cooperative/regret.py (hunk 7 above)
Comment (furadnik) [reply incomplete_cooperative/regret.py:136]: Sorry, but `meta_values` is not descriptive at all, I'm keeping this, because I wouldn't be able to read the code myself if I changed this :D

File: incomplete_cooperative/regret.py (hunk 2 above)
Comment (furadnik) [reply incomplete_cooperative/regret.py]: Wtf why is it dan

Comment (DavidSych) [reply incomplete_cooperative/regret.py]: Not really

File: incomplete_cooperative/regret.py (hunk 3 above)
Comment (furadnik) [reply incomplete_cooperative/regret.py]: This seems weird. I don't see the `weight` in the RM+ lecture...

File: incomplete_cooperative/regret.py (hunk 8 above)
Comment (furadnik) [reply incomplete_cooperative/regret.py:79]: I'll add `predictive` in a separate issue. (https://github.com/furadnik/IncompleteCooperative/issues/346)

File: incomplete_cooperative/regret.py (hunk 5 above)
Comment (furadnik) [reply incomplete_cooperative/regret.py]: I added this to the top of the function, as we're starting from zero (is that correct?)

File: incomplete_cooperative/regret.py (hunk 4 above)
Comment (furadnik) [reply incomplete_cooperative/regret.py:86]: Done.

File: incomplete_cooperative/regret.py (hunk 2 above)
Comment (furadnik) [reply incomplete_cooperative/regret.py]: Now :smile:

File: incomplete_cooperative/regret.py (hunk 1 above)
Comment (furadnik) [reply incomplete_cooperative/regret.py:111]: Changed. The `plus` and `predictive` should I think go to the `__init__`, as we don't really want to use the same regret to minimize oncewith `plus` `True`, and then once with `False`... Makes sense?

File: incomplete_cooperative/regret.py [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+        return positive_regret / positive_regret.sum()
+
+    def get_average_strategy(self, past_actions: Iterable[Coalition]) -> np.ndarray:
+        """Get average strategy, in terms of actual coalitions in the original game."""
+        metacoalition = self.get_metacoalition_id(past_actions)
+        rank = self.meta_id_to_rank[metacoalition]
+        # in `regret minimization` we only add the "previous" regret matching strategy, not the current one,
+        # so we need to add it here
+        average_strategy = self.cumulative_strategy[rank] + self.regret_matching_strategy(metacoalition)
+        average_strategy = average_strategy / average_strategy.sum()
```
Comment (furadnik) [inline incomplete_cooperative/regret.py:141]: is this correct?

Comment (DavidSych) [reply incomplete_cooperative/regret.py:140]: I don't see why you need to add the `regret_matching_strategy` here. It probably depends on when you call this, but I would expect this method to give me just the normalized `self.cumulative_strategy`.

File: incomplete_cooperative/regret.py [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+        # translate from player_ids to coalition ids
+        return average_strategy[self.coalitions_to_player_ids] * (self.coalitions_to_player_ids > -0.5)
```
Comment (DavidSych) [inline incomplete_cooperative/regret.py:143]: Why is there the last part? And wouldn't it mess with the normalization?

File: incomplete_cooperative/regret.py (hunk 7 above)
Comment (DavidSych) [reply incomplete_cooperative/regret.py:158]: Sure, I just wanted to push the meta in there somewhere to make it clear it is over meta-coalitions.

File: incomplete_cooperative/regret.py (hunk 11 above)
Comment (furadnik) [reply incomplete_cooperative/regret.py:140]: Here's how I imagine it:
1. We "train" the minimizer, essentially repeatedly calling the `regret_min_iteration` method. This always updates the cumulative losses and the cumulative strategy, and _then_ it updates the cumulative regret.
2. We ask for the current average strategy. But the current cumulative regret wasn't yet used to update the cumulative strategy (because in `regret_min_iteration`, it was updated _after_ the cumulative strategy).

So basically, I feel like if we don't add the regret matching strategy, we have one "update of regret" we're not using.
This is my intuition, is it correct? Also, is that how it is done in the literature?

**Comment by furadnik** [comment]
Also, please, when you feel like you're done with a conversation, mark it as "resolved", so it's clear what still requires work :)

File: incomplete_cooperative/regret.py (hunk 12 above)
Comment (furadnik) [reply incomplete_cooperative/regret.py:143]: Right, that's a bit of a hack. I'll add a comment about it.

Essentially, in the average strategy we store, we remove the coalitions which are already a part of the K_0.
In the output strategy, we're expecting them to be there again. This line fixes that by indexing with the right ids, where we have positions even for the coalitions from K_0, and they're filled with -1s. So what we get is a re-ordered array, and for every element not in the original array, the element at position -1 is taken. So if we then set the positions where the index is -1 (that's what the multiplication is doing), we set those positions to 0, which is what we want.

File: incomplete_cooperative/regret.py (hunk 3 above)
Comment (DavidSych) [reply incomplete_cooperative/regret.py]: The `weight` doesn't have to be there and it is still guaranteed to converge. But this should still have convergence guarantees and it is empirically faster. In fact, RM+ empirically converges in last-iterate, so we might not care about averaging. 

But one thing we should care about is that the cumulative strategy should be multiplied by the reach probability of this node. And that might be annoying to do here...

File: incomplete_cooperative/regret.py (hunk 11 above)
Comment (DavidSych) [reply incomplete_cooperative/regret.py]: I think it is technically correct, but I really don't like this design. For example, we can later add a second function which needs the `average_strategy`, and this extra evaluation will ruin the convergence. Accumulating regret should be done in the main loop.

Comment (furadnik) [reply incomplete_cooperative/regret.py]: Ok. Changed.

File: incomplete_cooperative/regret.py (hunk 12 above)
Comment (furadnik) [reply incomplete_cooperative/regret.py:149]: The values in average_strategy are normalized, and the other values are zero, so it wouldn't mess with the normalization. But just to be sure, I moved the normalization to a later stage.

File: incomplete_cooperative/regret.py (hunk 3 above)
Comment (furadnik) [reply incomplete_cooperative/regret.py]: What's "reach probability"? That means like under the current policy, ie the current average strategy? That would be annoying lol

File: incomplete_cooperative/regret.py [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+        # regret minimizer for every internal node -- except for the bottom layer of game tree
+        self.number_of_regret_minimizers = coalitions_up_to(self.number_of_coalitions, limit_of_revealed - 1)
+
+        self.coalitions_to_player_ids = get_coalition_player_id_map(number_of_players)
+        self.cumulative_regret = np.zeros((self.number_of_regret_minimizers, self.number_of_coalitions), dtype=RMValue)
+        self.cumulative_strategy = np.zeros((self.number_of_regret_minimizers, self.number_of_coalitions),
+                                            dtype=RMValue)
+        self.iteration = 0
+
+    def save(self, path: Path) -> None:
```
Comment (furadnik) [inline incomplete_cooperative/regret.py:112]: Finally implemented save and load methods. I decided to store the np arrays using `np.save`, as I think it might be more optimized. The rest is stored as a json file next to the stored np arrays. This means that it doesn't create one file, but rather a directory of 3 files, but I don't think that is a problem. What do you think?

File: incomplete_cooperative/regret.py (hunk 3 above)
Comment (DavidSych) [reply incomplete_cooperative/regret.py]: Under current policy. The counterfactual value just sums rewards from the bottom of the tree, but the addition to the cumulative strategy (the thing which if you normalize you get the average strategy with convergence guarantees) must be multiplied by the reach probability under current strategy. 

The idea is that the empirical average converges - so if I choose a random _time_ t<T, I should get a strategy in the tree which converges. This is not the same as picking the state. Assume T=2 and budget two. For t=1 you always pick a large coalition at first, and in the state where anything is revealed, you pick something small. Then for t=2, you pick small at first, and large later. Now what is the average prob of choosing small, given you already chose small (i.e. the second budget step). The average there is 50/50, but that is not what the empirical average would give us. Anytime we would actually visit the "I selected small at first" state we picked large.

File: incomplete_cooperative/regret.py (hunk 13 above)
Comment (DavidSych) [reply incomplete_cooperative/regret.py:112]: For 2D arrays, I think `np.savetxt` uses less memory. But it should be fine regardless. Feel free to resolve this conversation.

Comment (furadnik) [reply incomplete_cooperative/regret.py:112]: Good to know. Leaving it as is now, and if it is horrendous, I will change it.

File: incomplete_cooperative/regret.py (hunk 3 above)
Comment (furadnik) [reply incomplete_cooperative/regret.py]: Added.

Comment (DavidSych) [reply incomplete_cooperative/regret.py]: If I understand the code right, you are computing the reach using the average strategy. This is wrong, you should use the current strategy. So when going down the tree, you should just call `self.regret_matching_strategy(int(next_metacoalition))` (I think), and update the reach probability.

Comment (furadnik) [reply incomplete_cooperative/regret.py]: Fixed.
